### ISSUE TITLE
Names support for IRC

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3275]
|<img alt="Leonidas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Leonidas reported|
|-|-|
|Trac ID|trac#3275|
|Type|enhancement|
|Created|2008-06-04 11:58:47Z|
|Branch|https://github.com/twisted/twisted/tree/add-irc-names-support-3275|

Hi,

at the moment Twisted Words does not support the `NAMREPLY` message that is needed to get the users that are inside a channel which makes it hard to create simple clients.

In the patch attached I have added a handler for `irc_RPL_NAMREPLY`, `names` to trigger the `NAMES` command and `channelNames` which can be overridden in an easy way to implement own clients.

I am not sure what the best way of handling ops, voices etc is. My patch leaves the `@%+!` alone there is also some other implementation of `irc_RPL_NAMREPLY` [pasted here](http://dpaste.com/54756/) which might or might not be better. A problem of that one is, that some networks have more than just `@%+` user modes.

Attachments:

* [names-alternate.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/139/139171fad78cfb750862d199c90193938cef0091/d627949f3a210672bc4083d1f1544513266d7cbd.diff) (814 bytes) - added by Leonidas on 2008-06-04 12:02:39Z - Alternate implementation of NAMREPLY, as linked.
* [words-irc-names.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/139/139171fad78cfb750862d199c90193938cef0091/d24632e6ee3f68d5dd5497444e2c8d4efffd01a0.diff) (12429 bytes) - added by Leonidas on 2008-06-10 09:59:53Z - Ugly but supposedly feature-complete patch for NAMES support.
* [names-support-patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/139/139171fad78cfb750862d199c90193938cef0091/57647dc0076236ebacbef46193f3c99fce86bbf3.patch) (8063 bytes) - added by nlloyd on 2013-09-05 19:16:34Z - Patch without deferreds and including unit tests
* [names-support-patch.v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/139/139171fad78cfb750862d199c90193938cef0091/51c300198ed90d1b6b214a334e3b8bcce1078910.patch) (11338 bytes) - added by nlloyd on 2013-09-11 01:09:38Z - 
* [3275-styleAndTestFixes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/139/139171fad78cfb750862d199c90193938cef0091/f5f2d53c1950642b79043ca8dc652b893f7b9af5.patch) (15417 bytes) - added by nlloyd on 2013-09-30 23:01:10Z - 
* [3275-newsFile.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/139/139171fad78cfb750862d199c90193938cef0091/3cd7a830b7c99dbe46a5fdb80627d309aa581047.patch) (360 bytes) - added by nlloyd on 2013-09-30 23:10:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3275 3275
type__enhancement enhancement
reporter__Leonidas Leonidas
priority__normal normal
milestone__ 
branch__branches_add_irc_names_support_3275 branches/add-irc-names-support-3275
branch_author__Leonidas__nlloyd Leonidas, nlloyd
status__new new
resolution__None None
component__words words
keywords__ 
time__1212580727000000 1212580727000000
changetime__1389806087000000 1389806087000000
version__None None
owner__nlloyd nlloyd
cc__Wolf
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to Leonidas|
|-|-|

This needs unit tests.  Also, multiple NAMREPLY responses are possible.  You need to wait for ENDOFNAMES.

[sandbox/exarkun/irc2.py irc2.py](sandbox/exarkun/irc2.py irc2.py) might be worth a look.  There's a test module next to it as well.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Thanks a lot for your patch.  We definitely need more volunteers to help out with `twisted.words` maintenance.  However, it definitely needs some work.  First, a bit of an apology: twisted.words is really old code, and it doesn't set a great example for the standard of code quality required of new changes.  You should definitely read the [twisted development wiki page](https://twistedmatrix.com/trac/wiki/TwistedDevelopment) and all of the pages it links to to understand these standards.

As exarkun says, this code first and foremost needs unit tests.  There are other issues as well.

  * `names-alternate.diff` includes a reference to a function which is not defined anywhere, `parse_user`.
  * Don't worry about handling all possible modes in one ticket.  Represent the user as a simple data structure; a name and set of modes attached to the IRCClient object.  More features can be added as individual additions to that object, rather than trying to add a lot of stuff at once.  (One rather obvious one would be a method to send a message to that user, but future modes could each be handled that way, as well as different ways that certain modes are handled by different networks.)
  * As a minor issue, now we tend to format docstrings
```
#!py
"""
like this, now
"""
```
```
#!py
"""not like this."""
```
  * It would be nice to reduce the level of subclassing required by `IRCClient`.  This could be done by firing a deferred with a list of users when `ENDOFNAMES` is reached.  `irc2.py` provides a nice example of how this can be done without incurring race conditions.  To handle large lists of names, you might also allow the API user to pass in an object that receives each user object and then a "done" notification.  Again, I'm sorry that the existing IRC code doesn't provide a terribly good example here - the other methods all require too much subclassing to handle too.

Thanks again for looking at twisted.words.  Good luck with your modifications!

**Comment by twisted-trac** [comment]
|<img alt="Leonidas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Leonidas commented|
|-|-|

Replying to [glyph](#note_2):

> As exarkun says, this code first and foremost needs unit tests.  There are other issues as well.

I'm going to supply these. I need to take a look how unit tests in Twisted look like and it shouldn't be a problem.

>   * It would be nice to reduce the level of subclassing required by `IRCClient`.  This could be done by firing a deferred with a list of users when `ENDOFNAMES` is reached.  `irc2.py` provides a nice example of how this can be done without incurring race conditions.  To handle large lists of names, you might also allow the API user to pass in an object that receives each user object and then a "done" notification.  Again, I'm sorry that the existing IRC code doesn't provide a terribly good example here - the other methods all require too much subclassing to handle too.

Did I get it right? You propose instead of overwriting the callbacks to return deferreds? I have already started to do some modifications to `join()` and `names()` to return proper deferreds but while discussing it on IRC Taggnostr said it would be unwise to introduce yet another paradigm (subclassing and deferreds). While it is definitely possible to add some more deferreds without breaking compatibility, question is whether is is a good idea. Therefore I am asking people with  more experience.

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|Wolf commented|
|-|-|

I think that something like http://dpaste.com/55119/ could work.
(On the line 11 I hardcoded '@%+', but the server sends the modes that it accepts just after the connection (RAW 0003 in PREFIX), so they should be saved in a var and used here.)
irc_RPL_NAMREPLY saves all the nicks in a dict like {channelname: [('@', nick), ('+', nick2), (None, nick3), ...]}.
This function also receives the name of the server (prefix), my nick (params[0]) and the channel type (param[1]). The ENDOFNAMES receive the name of the server, my nick, but not the channel type, so if we want to send it to the namreply we should save it somewhere in order to use it when the ENDOFNAMES is called (we could ignore the servername and my nick here).
When irc_RPL_ENDOFNAMES is called the list and the channel name are sent to the API  function namreply and the channel is removed from the dict in order to prevent that further calls to NAMES will add other users to the others in the list (afaik there isn't a STARTOFNAMES so we can't use that for empty the list).

If you think that this is ok I can improve it, add the tests and submit the patch.
We also need something to parse the RAW 003 and save the prefix and the other things somewhere.

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|Wolf commented|
|-|-|

Replying to [Wolf](#note_5):
> (On the line 11 I hardcoded '@%+', but the server sends the modes that it accepts just after the connection (RAW 0003 in PREFIX), so they should be saved in a var and used here.)
> We also need something to parse the RAW 003 and save the prefix and the other things somewhere.
See [#3285](https://github.com/twisted/twisted/issues/3285)

**Comment by twisted-trac** [comment]
|<img alt="Leonidas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|Leonidas commented|
|-|-|

I am working on this in this bzr branch: <https://code.launchpad.net/~leonidas/twisted/improved-names-3275>. I'll keep this ticket updated as I make progress. So far, I got some deferreds working and write some tests. More will follow.

**Comment by twisted-trac** [comment]
|<img alt="Leonidas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|Leonidas commented|
|-|-|

  Ugly but supposedly feature-complete patch for NAMES support.

For everybody interested, I created a patch out of my branch. While the code is not yet ready for review, it more or less shows in which direction this is going. `names()` both calls `channelNames()` and returns a deferred which is fired once the request is finished.

**Comment by twisted-trac** [comment]
|<img alt="foobarbecue's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|foobarbecue commented|
|-|-|

Leonidas, what's the status of this? I'd really like to use NAMES in my irc bot!

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|nlloyd commented|
|-|-|

2 years later... what is the status of this?  Any chance of this getting into the next release?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Replying to [nlloyd](#note_10):
> 2 years later... what is the status of this?

Looks like Leonidas stopped working on his branch quite a while ago.  Maybe you could continue their work?

> Any chance of this getting into the next release?

There's always a chance!  Not into the release currently being prepared, since a pre-release has already gone out, but for the next one, sure!  All that we need is for somebody to fix the bug and somebody to do the code review.  Do you feel like taking on Leonidas's patch?

**Comment by twisted-trac** [comment]
|<img alt="Leonidas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|Leonidas commented|
|-|-|

Well yes, I stopped working on it 4 years ago now (was already wondering why nlloyd said 2 years). I'll be happy if you take over and rebase it on top of the current release and get it into a release. Good luck, if you have questions, I'm around.

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|nlloyd commented|
|-|-|

I have integrated the changes to support names into my project's copy of the Twisted library.  Once I have worked out a few of the kinks (and written more tests) I will happily submit a patch for the next (or next-next) release!

And the 2 years reference was to the last comment before my comment on this post.

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|nlloyd removed owner|
|-|-|

I took the patches above and some time ago integrated them into a Sublime Text 2 plugin I have been developing.  I pulled out the deferreds to keep things simpler and stay consistent with the rest of the IRCClient implementation.  I also included unit test coverage of the new functionality.

Unless I missed something this patch is ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set owner to @habnabit|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_16"></a>|@habnabit set owner to nlloyd|
|-|-|

Hi! Thank you for the patch! Just a few things that should be changed:

 1. `twisted.words.protocols.irc`
     1. Instead of hardcoding `NICK_PREFIXES` (which is, sadly, incomplete), use
        IRCClient's ISUPPORT functionality to learn which nickname prefixes are in
        use. 
     2. Your logic on lines lines 1447-1452 could be simplified into just an
        if/else, considering that you end up doing a for loop over `channels`
        anyway.
     3. Minor nit, but lines 1881-1882 could be condensed into
        `self._namreply.setdefault(channel, []).extend(users)`.
     4. Line 1893 seems unnecessary.
     5. Line 1894 should use `x not in y` instead of `not x in y`.
     6. On line 2474, did you mean "NAMES replies"?
 2. `twisted.words.test.test_irc`
     1. You seem to have accidentally removed a necessary newline.
     2. Instead of making one test with many assertions, `test_names` should be
        split up into multiple test methods. It looks like this should be six
        tests; two for each section delimited by `clear()` calls (one for
        calling `.names()` and one for calling `.irc_RPL_*`).
        
By the way, here's what you get out from the ISUPPORT functionality, assuming
`proto` is an IRCClient instance:

```
>>> proto.supported.getFeature('PREFIX')
{'o': ('@', 0), 'v': ('+', 1)}
```

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|nlloyd set owner to @habnabit|
|-|-|

The latest patch should address these issues but please let me know what, if anything, I missed.  A few comments are below as well.

Replying to [habnabit](#note_16):
> Hi! Thank you for the patch! Just a few things that should be changed:
> 
>  1. `twisted.words.protocols.irc`
>      1. Instead of hardcoding `NICK_PREFIXES` (which is, sadly, incomplete), use
>         IRCClient's ISUPPORT functionality to learn which nickname prefixes are in
>         use. 
>      2. Your logic on lines lines 1447-1452 could be simplified into just an
>         if/else, considering that you end up doing a for loop over `channels`
>         anyway.
>      3. Minor nit, but lines 1881-1882 could be condensed into
>         `self._namreply.setdefault(channel, []).extend(users)`.
>      4. Line 1893 seems unnecessary.
>      5. Line 1894 should use `x not in y` instead of `not x in y`.
>      6. On line 2474, did you mean "NAMES replies"?
>  2. `twisted.words.test.test_irc`
>      1. You seem to have accidentally removed a necessary newline.
**[nlloyd] I think I fixed this but double check please, and if I didn't could you point out the particular line?**
>      2. Instead of making one test with many assertions, `test_names` should be
>         split up into multiple test methods. It looks like this should be six
>         tests; two for each section delimited by `clear()` calls (one for
>         calling `.names()` and one for calling `.irc_RPL_*`).
""[nlloyd] Since names() and irc_RPL_*() calls are a matched set I included the end-to-end with mock replies from the server in the test.  I did, however, break up the giant test into several distinct scenarios.""
>         
> By the way, here's what you get out from the ISUPPORT functionality, assuming
> `proto` is an IRCClient instance:
> 
> ```
> >>> proto.supported.getFeature('PREFIX')
> {'o': ('@', 0), 'v': ('+', 1)}
> ```
> 
> Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_18"></a>|@habnabit commented|
|-|-|

(In [39947]) Branching to 'add-irc-names-support-3275'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_19"></a>|@habnabit commented|
|-|-|

(In [39948]) Apply nlloyd's patch.

Refs [#3275](https://github.com/twisted/twisted/issues/3275).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set owner to nlloyd|
|-|-|

Thanks for your patch, nlloyd. Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/add-irc-names-support-3275).

 1. This needs a news file
 1. There's a list of twistedchecker warnings that need to be fixed: http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1310/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
 1. There are 4 tests failing, see bottom of http://buildbot.twistedmatrix.com/builders/py-without-modules/builds/1748/steps/trial/logs/stdio

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|nlloyd set owner to @thijstriemstra|
|-|-|

I have yet to tackle the news file but I did address the twistedchecker and unit test issues.  Since apparently I cannot commit to the branch associated with this ticket (should I be able to?) there is a new patch file for the changes (3275-styleAndTestFixes.patch).

Replying to [thijs](#note_20):
> Thanks for your patch, nlloyd. Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/add-irc-names-support-3275).
> 
>  1. This needs a news file
>  1. There's a list of twistedchecker warnings that need to be fixed: http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1310/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
>  1. There are 4 tests failing, see bottom of http://buildbot.twistedmatrix.com/builders/py-without-modules/builds/1748/steps/trial/logs/stdio

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|nlloyd commented|
|-|-|

And now I have the news file patch attached :-)

Replying to [nlloyd](#note_21):
> I have yet to tackle the news file but I did address the twistedchecker and unit test issues.  Since apparently I cannot commit to the branch associated with this ticket (should I be able to?) there is a new patch file for the changes (3275-styleAndTestFixes.patch).
> 
> Replying to [thijs](#note_20):
> > Thanks for your patch, nlloyd. Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/add-irc-names-support-3275).
> > 
> >  1. This needs a news file
> >  1. There's a list of twistedchecker warnings that need to be fixed: http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1310/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
> >  1. There are 4 tests failing, see bottom of http://buildbot.twistedmatrix.com/builders/py-without-modules/builds/1748/steps/trial/logs/stdio

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|nlloyd set owner to @habnabit|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="nlloyd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|nlloyd set owner to @thijstriemstra|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_27"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_28"></a>|@wallrj commented|
|-|-|

(In [41251]) Apply 3275-styleAndTestFixes.patch  from nlloyd. Refs [#3275](https://github.com/twisted/twisted/issues/3275)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_29"></a>|@wallrj commented|
|-|-|

(In [41252]) Apply 3275-newsFile.patch from nlloyd. Refs [#3275](https://github.com/twisted/twisted/issues/3275)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_30"></a>|@wallrj commented|
|-|-|

Forced build:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/add-irc-names-support-3275

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_31"></a>|@wallrj set owner to nlloyd<br>@wallrj set status to `new`|
|-|-|

Thanks nlloyd, Leonidas and everyone who's contributed to this ticket.

I'm no IRC expert, but the changes in this branch look pretty sane to me.

Most of my comments are nitpicks and cosmetic problems. Some of the them need
addressing though before this branch can land.
I've marked the points that *must* be addressed with a *

Notes:
 1. Applied the latest two patches from nlloyd
 2. Merges cleanly to trunk

Points:
 1. Build Results
    1. https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/add-irc-names-support-3275
    2. Unit tests all pass
    3. But there are a few twistedchecker warnings which need addressing

 2. [branches/add-irc-names-support-3275/twisted/words/test/test_irc.py](branches/add-irc-names-support-3275/twisted/words/test/test_irc.py)
    1. "2007	class SemiImplClient"
       1. I'd prefer a more readable class name. Maybe "PartialClient" or at least "SemiImplementedClient"
       2. *Too many blank lines before first method. Should be 0 or 1.
       3. It would be clearer if the test specific methods were private. To
          distinguish them from the IRCClient methods which are being
          overridden.
       4. userNamesByChannel would be better as a private variable for the same reason.
       5. *Is it necessary to call transport.clear and protocol.clear at the end
          of test methods? These variables will be recreated by setUp before
          each test method gets run.
       6. test_names_no_channels
          1. Instead of calling rstrip on the transport value, it might be better
             to add the expected newline to the expected value. Then you are
             testing exactly what bytes the protocol writes to the transport.
          2. And, to me, it would be clearer to do assertIn(oneChannel,
             self.protocol._userNamesByChannel), then you could ditch the
             hasChannel convenience method and simplify the test fake. It would
             also give clearer output in the case of a test failure.
       7. test_names_one_channel
          1. *The docstring is identical to the test above. Needs updating.
          2. Some of same comments as for no_channels test
       8. test_names_many_channels
          1. *The docstring is identical to the test above. Needs updating.
          2. Some of same comments as for no_channels test
          3. *"for channel in manyChannels" loops in tests make it difficult to
             see multiple failures (if they occur).
             1. Better to call assertEqual once with an expected list and an actual
                list of results.
       9. test_names_many_channels_many_user_types
          1. *The docstring is identical to the test above. Needs updating.
          2. Some of same comments as for no_channels test

 3. [branches/add-irc-names-support-3275/twisted/words/protocols/irc.py](branches/add-irc-names-support-3275/twisted/words/protocols/irc.py)
    1. *Line 1 is incorrectly indented -- a typo I guess
    2. "642                 'PREFIX': self._parsePrefixParam('(qaovh)~&@+%'), "
       1. I struggled to find a definitive reference to the new mode characters
          here.
       2. It would be nice to have such a link if one exists.
    3. *"1093        # cache "
       1. The comment should be moved to the docstring as an @ivar. All instance
          variables (public and private) must be documented.
       2. Also comments should start with a capital letter.
       3. http://nedbatchelder.com/blog/201401/comments_should_be_sentences.html
       4. https://bazaar.launchpad.net/~twisted-dev/twistedchecker/trunk/view/head:/twistedchecker/checkers/comment.py#L23
    4. 1276 "the name of the channel where the users are in. "
       1. Would sound better as "The name of the channel containing the users."
       2. *Also these docstrings should be sentence case.
       3. Maybe add an adjacent @type documentation for completeness
    5. "1438        def names"
       1. *Should be two blank lines between methods
       2. *"`channelNames`" should be L{channelNames} which will produce a link in the rendered documentation.
       3. *"@type channels: C{str}" is actually a L{list} or L{str}
       4. *Typo "or or"
       5. And this would probably sound better "The name of the channel or channels for which to request
          username lists from the server."
    6. "1874        def irc_RPL_NAMREPLY"
       1. "returned as answer to" -> "returned in answer to"
       2. *C{str}  should be L{str}
       3. *C{Array} should be L{list}
       4. *"parameters for the RPL_NAMREPLY message" should be sentence case and is missing a full stop.
       5. Also you talk about the third and fourth entries, but what about the first and second.
       6. The comment "# cache nickname prefixes..." looks like it would be
          useful info to include in the docstring. May as well move it there."
       7. Might be simpler as  ''.join(v[0] for v in prefixes.itervalues())
    7. "1900        def irc_RPL_ENDOFNAMES"
       1. Many of the same issues as above
       2. This sentence was hard to read -- "It gathers one, or all depending
          on the NAMES request, channel names lists gathered from
          RPL_NAMREPLY responses."
    8. "2510            # container for NAMES replies "
       1. *This documentation should be moved to the class docstring as an @ivar
          parameter.


Thanks again for your work. Please answer and address at least the highlighted
points above and resubmit for another (hopefully final) review.  I'm rwall
in #twisted-dev if you're unsure of, or disagree with any my points.
